File: course/src/11_interfaces/DuckTyping.java
interface Duck {
    void quack();
    void swim();
}

class RealDuck implements Duck {
    @Override
    public void quack() {
        System.out.println("Quack!");
    }
    @Override
    public void swim() {
        System.out.println("Swimming in the pond!");
    }
}

class Dog {
    public void quack() {
        System.out.println("Bark like a duck!");
    }
    public void swim() {
        System.out.println("Swimming like a dog!");
    }
}

public class Main {
    public static void makeItQuackAndSwim(Duck duck) {
        duck.quack();
        duck.swim();
    }

    public static void main(String[] args) {
        RealDuck duck = new RealDuck();
        makeItQuackAndSwim(duck); // Works, RealDuck implements Duck

        // Dog dog = new Dog();
        // makeItQuackAndSwim(dog); // Compile-time error: Dog is not a Duck
    }
}

File: course/src/02_abstract_data_types/11_interfaces/ducktyping.py
class RealDuck:
    def quack(self):
        print("Quack!")

    def swim(self):
        print("Swimming in the pond!")

class RubberDuck:
    def quack(self):
        print("Squeak!")

    def swim(self):
        print("Floating in the bathtub!")

class Dog:
    def quack(self):
        print("I can't quack like a duck, but I'll bark instead!")

    def swim(self):
        print("Swimming like a dog!")

def make_it_quack_and_swim(duck):
    # Duck typing: no need to check the type of 'duck', just call methods
    duck.quack()
    duck.swim()

# Any object with quack() and swim() methods will work
make_it_quack_and_swim(RealDuck())
make_it_quack_and_swim(RubberDuck())
make_it_quack_and_swim(Dog())

